test/cpp/src/200-rob-scheduling.cc

```diff
@@ -130,3 +130,144 @@ SCENARIO("The scheduler can detect RAW hazards") {
   }
 }
 
+SCENARIO("The scheduler handles WAW hazards") {
+  GIVEN("A ROB with a WAW hazard followed by a read from that register") {
+    constexpr unsigned schedule_width = 128;
+    constexpr unsigned schedule_latency = 1;
+    constexpr unsigned execute_width = 3;
+    constexpr unsigned execute_latency = 3;
+
+    do_nothing_MRC mock_L1I, mock_L1D;
+    O3_CPU uut{champsim::core_builder{}
+      .schedule_width(champsim::bandwidth::maximum_type{schedule_width})
+      .schedule_latency(schedule_latency)
+      .execute_latency(execute_latency)
+      .execute_width(champsim::bandwidth::maximum_type{execute_width})
+      .retire_width(champsim::bandwidth::maximum_type{execute_width})
+      .fetch_queues(&mock_L1I.queues)
+      .data_queues(&mock_L1D.queues)
+    };
+
+    uut.ROB.push_back(champsim::test::instruction_with_ip(1));
+    uut.ROB[0].instr_id = 1;
+    uut.ROB[0].destination_registers.push_back(5);
+    uut.ROB.push_back(champsim::test::instruction_with_ip(2));
+    uut.ROB[1].instr_id = 2;
+    uut.ROB[1].destination_registers.push_back(5);
+    uut.ROB.push_back(champsim::test::instruction_with_ip(3));
+    uut.ROB[2].instr_id = 3;
+    uut.ROB[2].source_registers.push_back(5);
+    for (auto &instr : uut.ROB)
+      instr.ready_time = champsim::chrono::clock::time_point{};
+
+    WHEN("The first two instructions are in flight") {
+      uut.ROB[0].scheduled = false;
+      uut.ROB[1].scheduled = false;
+      uut.ROB[2].scheduled = false;
+      // Schedule
+      for (auto op : std::array<champsim::operable*,3>{{&uut, &mock_L1I, &mock_L1D}})
+        op->_operate();
+      REQUIRE(uut.ROB[2].num_reg_dependent == 1);
+      // Execute
+      for (auto op : std::array<champsim::operable*,3>{{&uut, &mock_L1I, &mock_L1D}})
+        op->_operate();
+
+      THEN("The third instruction does not execute") {
+        REQUIRE(uut.ROB[0].executed == true);
+        REQUIRE(uut.ROB[1].executed == true);
+        REQUIRE(uut.ROB[2].executed == false);
+        REQUIRE(uut.ROB[2].num_reg_dependent == 1);
```

Comment:
I am somewhat concerned with examining the internals of `O3_CPU` and `ooo_model_instr` in this way. This test works and accomplishes what we want to do, but this test will break on the day when we modify the internals of either one of those classes.

Reply:
It's difficult, since to ensure for instructions A,B,C that B depends on C and that this dependency is respected in the scheduler, without the internals, we can, at best, confirm that 3 instructions retire in X cycles rather than X+1 cycles, but not draw any conclusions about how the instructions were scheduled.